File: src/js/main/commands/RichJson_ref.js
/*
    references the given member's value (any)
*/

import {isJsonObject} from "../helper/RichJsonHelper.js";
import {__RICH_JSON_COMMAND_PATH_DELIMITER} from "../core/RichJson.js";

export function __executeRefCommand(parser, context) {
    if (context.currentMember === "") {
        return context.root;
    }
    let prevMember = context.root;
    let currentAddress = context.currentAddress;
    let refs = context.currentMember.split(__RICH_JSON_COMMAND_PATH_DELIMITER);
    let ref = undefined;

    for (let i = 0; i < refs.length; ++i) {
        ref = refs[i];
        context.currentName = ref;
        if (isJsonObject(prevMember)) {
            if (Object.hasOwn(prevMember, ref)) {
                context.currentMember = prevMember[ref];
            } else {
                throw (`Member '${ref}' in '${parser.cache.resolveAddress(prevMember)}' does not exist`);
            }
        }
        context.currentAddress = isJsonObject(context.currentMember) || Array.isArray(context.currentMember)
            ? parser.cache.resolveAddress(context.currentMember)
            : parser.cache.resolveAddress(prevMember) + "_" + ref
        ;
        context.currentMember = parser.__parseRichJsonInMember();
        context.currentPath.push(ref);
        prevMember = context.currentMember;
    }
    context.currentAddress = currentAddress;
    context.currentPath.splice(context.currentPath.length - refs.length, context.currentPath.length);
    return context.currentMember;
}

File: src/python/main/commands/rich_json_ref.py
"""
References the given member's value (any)
"""
from ..helper.rich_json_helper import concat_strings, is_json_object
from ..core.rich_json_constants import _RICH_JSON_COMMAND_PATH_DELIMITER

def _execute_ref_command(parser, context):
    if context.current_member == "":
        return context.root

    prev_member = context.root
    current_address = context.current_address
    refs = context.current_member.split(_RICH_JSON_COMMAND_PATH_DELIMITER)

    for ref in refs:
        if is_json_object(prev_member):
            if ref in prev_member:
                context.current_member = prev_member[ref]
            else:
                raise ValueError(f"Member '{ref}' in '{parser.cache.resolve_address(prev_member)}' does not exist")

        if is_json_object(context.current_member) or isinstance(context.current_member, list):
            context.current_address = parser.cache.resolve_address(context.current_member)
        else:
            context.current_address = concat_strings(parser.cache.resolve_address(prev_member), "_", ref)

        context.current_member = parser._parse_rich_json_in_member(context)
        prev_member = context.current_member

    context.current_address = current_address
    return context.current_member